packages/core/docs/api/wallet.md

```diff
@@ -293,3 +293,36 @@ Signs one or more transactions before sending to the network. The user must be s
   });
 })();
 ```
+
+### `.signMessage(params)`
+
+**Parameters**
+- `params` (`object`)
+  - `message` (`string`): The message that wants to be transmitted.
+  - `recipient` (`string`): The recipient to whom the message is destined (e.g. "alice.near" or "myapp.com").
+  - `nonce` (`Buffer`): A nonce that uniquely identifies this instance of the message, denoted as a 32 bytes array (a fixed `Buffer` in JS/TS).
+  - `callbackUrl` (`string?`): Optional, applicable to browser wallets (e.g. MyNearWallet). The URL to call after the signing process. Defaults to `window.location.href`.
+  - `state` (`string?`): Optional, applicable to browser wallets (e.g. MyNearWallet). A state for authentication purposes.
+
+**Returns**
+- `Promise<void | SignedMessage>`: Browser wallets won't return the signing outcome as they may need to redirect for signing. For MyNearWallet the outcome is passed to the callback url.
+
+**Description**
+
+Allows users to sign a message for a specific recipient using their NEAR account, based on the [NEP413](https://github.com/near/NEPs/pull/413).
+
+**Example**
+
+```ts
+// MyNearWallet
+(async () => {
+  const wallet = await selector.wallet("my-near-wallet");
+  const message = "test message for verification";
+  let nonceArray: Uint8Array = new Uint8Array(32);
+  nonceArray = crypto.getRandomValues(nonceArray);
+  const nonce = Buffer.from(nonceArray);
```

Comment:
If the internal implementation is in charge of the protocol security, and developers do not have any use for the `nonce`, I think it is fair to remove it from the public interface. What do you think?

Reply:
As a dApp developer, they could use `nonce` to verify the signature, right? Correct me if I'm wrong.

Reply:
Oh, well, validation is currently expected to be done manually by dApp developer. I think it would be great to have a higher-level API for dApp developers. This documentation example needs the validation to be featured

Reply:
Do you mean adding the following to the sample?
```js
import { verifySignature, verifyFullKeyBelongsToUser } from "@near-wallet-selector/core";

// ...

const verifiedSignature = verifySignature({
  message,
  nonce,
  recipient,
  publicKey: signedMessage.publicKey,
  signature: signedMessage.signature,
});
const verifiedFullKeyBelongsToUser = await verifyFullKeyBelongsToUser({
  publicKey: signedMessage.publicKey,
  accountId: signedMessage.accountId,
  network: selector.options.network,
});
const ok = verifiedSignature && verifiedFullKeyBelongsToUser
```

Both functions are now exported from `@near-wallet-selector/core`. Not sure if we really need a 'higher-level' API for dApp dev.

Reply:
Yep, that is what I missed in the example.

I like this low-level API, but I would still consider having an API that won't expect developers don't forget to do the checks. I would expect that once the message is returned, it is already verified. What would be the use case for dApp developers needing to intervene and need control over each stage and have access to a potentially maliciously constructed message?

Reply:
Fair point.
I'm thinking alternating it like so, where we have a second params (verify will be defaulted to true)
```js
const signedMessage = await wallet.signMessage({
  message,
  recipient,
}, {verify: true});
```

Or we just derived the verify flag from the availability of `nonce` field.
eg: 
```js
// This will verify the message
await wallet.signMessage({
  message,
  recipient,
})

// This will NOT verify the message
await wallet.signMessage({
  message,
  nonce,
  recipient,
})
```

Anyways, I believe the verification part should be the responsibility of Wallet builders?

Reply:
I kinda agree about making the verification step a bit more intentional from the side of the dApps, since they could potentially do their own (short-lived) caching of a valid public key per user, which would allow for instant actions on their apps instead of always doing an extra API call to check for account valid keys.

In the future, there could also be a standard text payload for simple things like "proof of account ownership" which use the same API, but don't require the user to intentionally complete the action in the wallet every single time (in the case of extension wallets, if it was done once per requesting and currently connected dApp, it could be done automatically each time afterwards using background script communication). To require each of these types of calls to still verify on the blockchain would make things like simply "liking a post" or whatnot, take a lot longer than they need to- deteriorating the user experience.